### PR TITLE
fix: per-concept resources in learning plan (#168)

### LINKED ISSUE #168: Per-concept resources in learning plan + AI generation on cards
# Feature: concept-level structure + AI generation for learning plan

> Draft as a GitHub issue body. Paste into a new issue when approved.

## Context

The learning-plan renderer on the assessment report page has a rendering bug: each phase produces N concept cards, but **every card shows the same resources** because resources live at the phase level while concepts are rendered per-row.

Bug site — `frontend/src/components/learning-plan/PlanTimeline.tsx:36-42`:

```tsx
{phase.concepts.map((concept, i) => (
  <ConceptCard
    key={concept}
    concept={concept}
    resources={phase.resources}  // ← every card gets the full phase list
    index={i}
  />
))}
```

The root cause is in the **backend schema**: `LearningPhaseOut` has `concepts: list[str]` and `resources: list[ResourceOut]` as two parallel, disconnected arrays (`backend/app/models/assessment_api.py:83-89`). There is no way for the frontend to know which resource belongs to which concept.

Beyond the bug fix, we want each concept card to let the user **generate AI content on demand** — both (a) study material and (b) practice questions — scoped to that specific concept. The existing codebase already has both generators, but they are coupled to the adaptive-assessment state machine and cannot be reused as-is.

### Single learning-plan path (post-#166)

`#166` deleted the standalone `POST /api/learning-plan` route and its duplicate models, prompts, and schemas. The codebase now has **one** learning-plan path:

- **Schema (LLM output)**: `PlanOutput` / `PlanPhaseOutput` / `PlanResourceOutput` in `backend/app/agents/schemas.py:49-75`
- **Graph state**: `LearningPlan` / `LearningPhase` / `Resource` in `backend/app/graph/state.py:119-137`
- **API wire**: `LearningPlanOut` / `LearningPhaseOut` / `ResourceOut` in `backend/app/models/assessment_api.py:77-95`
- **Generator**: `generate_plan` in `backend/app/agents/plan_generator.py:14-65`
- **Prompt**: `PLAN_GEN_PROMPT` — now the only prompt in `backend/app/prompts/plan_generator.py` (whole file is 15 lines)
- **Storage**: JSONB column `AssessmentResult.learning_plan` at `backend/app/db.py:87`
- **Mapper**: `build_learning_plan_out` at `backend/app/services/assessment_mappers.py:85-127` (has dict/model branches)
- **Markdown export**: `build_assessment_markdown` at `backend/app/routes/export_utils.py:86-113` (reads JSONB dict directly, not via mapper)
- **Frontend consumer**: `AssessmentReportResponse["learningPlan"]` rendered by `PlanTimeline` → `ConceptCard`

All five places must change in lockstep.

### JSONB back-compat constraint

`AssessmentResult.learning_plan` is stored as JSONB. Existing rows are written with the OLD shape (`concepts: list[str]`, `resources` at phase level). The mapper at `backend/app/services/assessment_mappers.py:90-111` already has a dict-read branch for JSONB. **The reshape must keep that dict branch able to read old rows without error** — otherwise previously completed assessments stop rendering after deploy.

---

## Proposed structure

Concepts become first-class objects with their own resources. Resources move off the phase level.

```python
# backend/app/models/assessment_api.py
class ConceptOut(CamelModel):
    key: str           # slug(name), stable within the phase
    name: str
    description: str
    resources: list[ResourceOut]

class LearningPhaseOut(CamelModel):
    phase_number: int
    title: str
    concepts: list[ConceptOut]   # was list[str]
    rationale: str
    estimated_hours: float
    # NOTE: phase-level `resources` removed
```

Why this shape:

- **Solves the bug cleanly** — the frontend no longer needs to know a resource↔concept mapping because it's in the data.
- **Unlocks the AI generation UX** — a `ConceptOut` carries everything a generation endpoint needs (`name`, `description`, `key` for React), without any DB persistence.
- **Minimal LLM cost** — the prompt just asks the model to attach 2-4 resources to each concept instead of to the phase. No new data, just a different grouping.
- **No DB migration** — JSONB back-compat is handled in the mapper.

### Why NOT other options

- *Just fix the rendering (frontend-only workaround).* Rejected — there is no way to correctly associate resources to concepts on the frontend because the association doesn't exist in the data. You'd end up rendering resources once per phase in a dedicated section, which is a regression vs. the current intent of the design.
- *Keep flat concepts, add a separate `concept_resources: dict[str, list[Resource]]` map.* Rejected — two parallel structures that must stay in sync are worse than one nested structure.
- *Persist concepts as DB rows with UUIDs.* Rejected — the learning plan is already ephemeral (stored inside `AssessmentResult.learning_plan` JSONB). Persisting concepts separately is speculative design for a use case we don't have.

---

## Implementation plan

Two PRs recommended. **PR 1** must land as one commit because the schema reshape and the frontend bug fix cannot be staged (generated types would be out of sync with runtime otherwise).

#### 1.1 Backend schema

Edit these files together:

- `backend/app/agents/schemas.py:49-75`
  - Add `PlanConceptOutput { key: str, name: str, description: str, resources: list[PlanResourceOutput] }` (next to `PlanResourceOutput` at line 49).
  - Change `PlanPhaseOutput.concepts: list[PlanConceptOutput]` (currently `list[str]` at line 62).
  - **Remove** `PlanPhaseOutput.resources` (currently line 66).

- `backend/app/graph/state.py:119-131`
  - Add `ConceptItem(CamelModel)` with the same shape as `PlanConceptOutput` (place near `Resource` at line 119).
  - Change `LearningPhase.concepts: list[ConceptItem]` (currently `list[str]` at line 128); remove `LearningPhase.resources` (line 130).

- `backend/app/models/assessment_api.py:77-89`
  - Add `ConceptOut(CamelModel)` (next to `ResourceOut` at line 77).
  - Change `LearningPhaseOut.concepts: list[ConceptOut]` (currently `list[str]` at line 86); remove `LearningPhaseOut.resources` (line 88).

- `backend/app/agents/plan_generator.py:45-57`
  - Adapt the LLM-output → state mapping: for each `PlanPhaseOutput`, iterate `p.concepts` and build `ConceptItem`s with their own `resources`. Drop the phase-level `resources = [Resource(...) for r in p.resources]` line (currently line 47).
  - Add a slug helper. Check for prior art first (`Grep "def slugify" backend/`); if none, inline: `re.sub(r"[^a-z0-9]+", "-", name.lower()).strip("-")`.

- `backend/app/prompts/plan_generator.py` (whole file, 15 lines) — rewrite `PLAN_GEN_PROMPT`:
  - Instruct the LLM: "Each concept within a phase must carry its own 2-4 learning resources (video, article, project, exercise). Do not emit phase-level resources."
  - Add a one-line example showing the nested shape.
  - `PLAN_GENERATOR_SYSTEM_PROMPT` was removed by #166; the file is now a single-purpose pipeline prompt.

#### 1.2 JSONB back-compat shim (critical)

`backend/app/services/assessment_mappers.py:85-127` — update `build_learning_plan_out`:

- **Dict branch (lines 90-111)**: detect shape. If `phase["concepts"]` is a list of strings (old shape), promote each to `ConceptOut(key=slug(name), name=name, description="", resources=[])` and ignore the stored `phase["resources"]`. Log a single info message so we know how many legacy rows are being read. (Old rows lose phase-level resources on render — acceptable because the duplication bug was the whole reason they looked the way they did. Call this out in the PR description.)
- **Model branch (lines 113-127)**: direct field map from `ConceptItem` to `ConceptOut`. Remove the phase-level `resources` list comprehension.

Add a test `test_build_learning_plan_out_legacy_dict_shape` that feeds an old-shape dict and asserts the mapper produces a valid `LearningPlanOut` without error.

#### 1.2b Markdown export (critical — previously listed as "audit")

`backend/app/routes/export_utils.py:86-113` reads the JSONB dict directly (not via the mapper) to render the `## Learning Plan` section. It currently iterates `phase.get("concepts", [])` as a list of strings (line 99-101) and `phase.get("resources", [])` as phase-level resources (line 102-112). After the reshape, the stored shape becomes nested concepts.

Update `build_assessment_markdown` to handle **both shapes**:

```python
concepts = phase.get("concepts", [])
if concepts and isinstance(concepts[0], dict):
    # New shape: each concept has its own name + resources
    for c in concepts:
        parts.append(f"\n#### {c.get('name', '')}")
        if c.get("description"):
            parts.append(c["description"])
        for r in c.get("resources", []):
            # existing resource rendering
else:
    # Legacy shape: list[str] + phase.resources
    if concepts:
        parts.append(f"\n**Concepts:** {', '.join(concepts)}\n")
    for r in phase.get("resources", []):
        # existing resource rendering
```

Add a test `test_build_assessment_markdown_new_plan_shape` to `backend/tests/test_export.py` covering the new shape.

#### 1.3 Regenerate OpenAPI

```bash
cd backend && python scripts/export-openapi.py
cd frontend && npm run generate-api   # or: make generate-api from repo root
```

Confirm `frontend/src/lib/generated/api-client/types.gen.ts` now contains `ConceptOut` and that `LearningPhaseOut.resources` is gone.

Add re-exports in `frontend/src/lib/types.ts` if needed.

#### 1.4 Frontend rendering (bug fix)

- `frontend/src/components/learning-plan/PlanTimeline.tsx:36-42` — map concepts as objects:
  ```tsx
  {phase.concepts.map((concept, i) => (
    <ConceptCard
      key={`${phase.phaseNumber}-${concept.key}`}
      concept={concept}
      index={i}
    />
  ))}
  ```

- `frontend/src/components/learning-plan/ConceptCard.tsx:25-91`:
  - Change `ConceptCardProps` to `{ concept: ConceptOut; index: number }`.
  - Read resources from `concept.resources` instead of a separate prop.
  - Render `concept.name` (drop the `concept.replace(/_/g, " ")...` formatting hack at line 44 — the backend now supplies a human-readable name).
  - Render `concept.description` under the heading if non-empty.

- `frontend/src/lib/demo/fixtures.ts:349` (`DEMO_LEARNING_PLAN`) — update the demo learning-plan fixture to the new shape. Resources currently inlined at lines 360, 373, 386, 398 need to move inside concepts.
- `frontend/e2e/fixtures/mock-data.ts:79` (`learningPlan:`) — update the e2e mock fixture similarly.
- Sanity check: `frontend/src/components/learning-plan/PlanHeader.tsx:15` uses `p.concepts.length` — this still works since a list's length is shape-agnostic.

#### 1.5 Tests touched in PR 1

- `backend/tests/test_structured_output.py:51,55,68,91` — update `PlanOutput` / `PlanPhaseOutput` construction rows.
- `backend/tests/test_assessment_mappers.py:153-163` — update `phase.concepts` / `phase.resources` rows.
- `backend/tests/test_state.py:141` — update `LearningPhase(...)` construction.
- `backend/tests/test_assessment_service.py:335` — update `LearningPhase(...)` construction.
- `backend/tests/test_assessment_routes.py:540` — update `LearningPhase(...)` construction.
- `backend/tests/test_export.py` — update markdown assertions for the new shape; add a test for the legacy-dict branch.
- *(Removed: `test_models_gap_learning.py` was deleted by #166.)*
- New: `backend/tests/test_plan_generator_concepts.py` — mock `ainvoke_structured` to return a `PlanOutput` with nested concepts; assert the resulting `LearningPlan` has no phase-level resources and each concept has its own list.
- New: `frontend/src/components/learning-plan/PlanTimeline.test.tsx` — regression test: given a phase with 3 concepts each with a **distinct** resource, assert each rendered `ConceptCard` shows only its own resource.
- Update: `frontend/src/components/learning-plan/ConceptCard.test.tsx` (if present) for the new prop shape.
- Audit: `frontend/src/app/learning-plan/page.test.tsx` — currently mocks `PlanTimeline` away, should still pass but re-check.

---

#### 2.1 New pure-function generators

Reasoning: `_generate_single_content` in `backend/app/agents/content_nodes.py:236-327` is tied to `PrioritizedGap`, `LearningObjective`, `TaxonomyIndex`, CLT params, and Bloom distance — none of which exist in the standalone "concept from a plan" context. `question_generator.generate_question` in `backend/app/agents/question_generator.py:214-263` is even more deeply coupled since #167: it now pulls `latest_evaluation`, `questions_on_current_topic`, `question_history`, `knowledge_graph`, `used_types`, and a newly-added `performance_signal` derived from a 10-helper chain (`build_performance_signal`, `_sanitize_evidence`, etc.). **Reusing either is harder than writing two small new functions.** The *output schemas* (`ContentGeneratorOutput`, `QuestionOutput` — `schemas.py:114-117`, `:17-27`) are plain and reusable.

Note: #167 updated `QuestionOutput.question_type` taxonomy to `conceptual | code | debugging | design | trade-off` (`schemas.py:26`). The new concept-question prompt must instruct the LLM to use this taxonomy for consistency with the rest of the system.

New files:

- `backend/app/agents/concept_content.py`:
  ```python
  async def generate_concept_material(
      concept_name: str,
      concept_description: str,
      phase_title: str,
      phase_rationale: str,
  ) -> ContentGeneratorOutput: ...

  async def generate_concept_questions(
      concept_name: str,
      concept_description: str,
      phase_title: str,
      count: int = 3,
  ) -> list[QuestionOutput]: ...
  ```
  Both call `ainvoke_structured(...)` via `app.services.ai`. No `AssessmentState`, no `LearningMaterialState`, no `PrioritizedGap`, no `TaxonomyIndex`.

- `backend/app/prompts/concept_content.py` — two new system prompts, modelled after `CONTENT_GENERATOR_SYSTEM_PROMPT` / `QUESTION_GEN_PROMPT` but without Bloom/CLT variables, performance signal, or evidence sanitization (none of those have analogues outside the assessment state machine). For questions, ask for `count` distinct questions spanning the post-#167 taxonomy (`conceptual`, `code`, `debugging`, `design`, `trade-off`) in one call; wrap in:
  ```python
  class ConceptQuestionsOutput(BaseModel):
      questions: list[QuestionOutput]
  ```
  so `ainvoke_structured` can return them in one shot.

#### 2.2 Routes and wire models

- `backend/app/models/concept_generation.py` (new):
  ```python
  class ConceptMaterialRequest(CamelModel):
      concept_key: str
      concept_name: str
      concept_description: str
      phase_title: str
      phase_rationale: str

  class ConceptQuestionsRequest(CamelModel):
      concept_key: str
      concept_name: str
      concept_description: str
      phase_title: str
      count: int = 3  # @field_validator: 1 <= count <= 10

  class QuestionOut(CamelModel):   # camelCase wire shape for QuestionOutput
      topic: str
      bloom_level: str
      text: str
      question_type: str

  class ConceptMaterialResponse(CamelModel):
      sections: list[ContentSectionOut]   # camelCase for ContentSectionOutput

  class ConceptQuestionsResponse(CamelModel):
      questions: list[QuestionOut]
  ```
  (Do **not** expose `ContentGeneratorOutput` / `QuestionOutput` directly — they are plain `BaseModel` for LLM use, not `CamelModel`.)

- `backend/app/routes/concept_generation.py` (new):
  - `POST /api/learning-plan/concept/materials` → `ConceptMaterialResponse`
  - `POST /api/learning-plan/concept/questions` → `ConceptQuestionsResponse`
  - Wrap in `api_key_scope(api_key)`; mirror error handling from any existing route (e.g. `routes/gap_analysis.py` or `routes/assessment.py`) — `logger.exception` + 500 `HTTPException`. (`routes/learning_plan.py` was deleted by #166, so don't look there.)

- `backend/app/main.py` — register `concept_generation.router` under `/api` (find the existing router registrations and add alongside).

#### 2.3 Regen OpenAPI and frontend client

```bash
cd backend && python scripts/export-openapi.py
cd frontend && npm run generate-api
```

#### 2.4 Frontend UI

- `frontend/src/lib/api.ts` — add two helpers: `generateConceptMaterial(req)` and `generateConceptQuestions(req)`. Standard `fetch` + `unwrap()` pattern — mirror an existing helper in the same file (e.g. `requestGapAnalysis` or the assessment helpers).

- `frontend/src/components/learning-plan/ConceptCard.tsx` — add two buttons under the resources list:
  - **"Study material"** — click → loading skeleton → inline `<ConceptMaterialPanel sections={...} />`
  - **"Practice questions"** — click → loading skeleton → inline `<ConceptQuestionsPanel questions={...} />`
  - Local state via `useState` for `{ loading, error, payload }` on each button.
  - A small "Regenerate" icon button to re-call the API after first fetch (cheap, good UX).
  - Ephemeral state only — no server-side caching; no localStorage; re-navigating away clears state. Called out as a v1 constraint; follow-up issue if users want persistence.

- New components: `ConceptMaterialPanel.tsx`, `ConceptQuestionsPanel.tsx` in `frontend/src/components/learning-plan/`. Presentational only. Render `section.type`-aware cards for materials (explanation / code_example / analogy / quiz) and a numbered list for questions (with `bloom_level` badge).

- `frontend/src/lib/types.ts` — re-export the new request/response types.

#### 2.5 Tests touched in PR 2

- New: `backend/tests/test_concept_generation_route.py` — follow the shape of any existing route test (e.g. `backend/tests/test_gap_analysis_route.py` or `test_assessment_routes.py`):
  - 422 on missing fields
  - 200 with mocked `ainvoke_structured` returning a valid `ContentGeneratorOutput` / `ConceptQuestionsOutput`
  - 500 on LLM exception
  - camelCase assertion on response body
  - `count` validator boundary cases (0, 1, 10, 11)
- New: `backend/tests/test_concept_content_agent.py` — unit tests for the two pure functions with mocked LLM.
- Update: `frontend/src/components/learning-plan/ConceptCard.test.tsx` — tests for the two buttons (loading state, success render, error state, regenerate).
- New: `frontend/e2e/learning-plan-generation.spec.ts` — Playwright journey: load plan → click "Practice questions" on concept 1 → assert panel shows 3 questions (mock the API via `page.route`).

---

## Deferred to follow-up issues (out of scope)

- **SSE streaming for generation endpoints.** The existing `SSEAdapter` (`backend/app/services/sse_adapter.py`) is coupled to assessment events, and the frontend has no shared SSE hook (parsing is inlined in `useAssessmentChat.ts:132-208`). For v1, non-streaming JSON is simpler, reuses error handling, and has good enough UX if the frontend shows a skeleton. Add streaming only if perceived latency becomes a real complaint.
- **Server-side caching of generated content.** The current `MaterialResult` table is session-scoped (`backend/app/db.py:109-130`, `UniqueConstraint("session_id", "concept_id")`) and doesn't fit an ephemeral learning-plan concept. Adding a separate cache is speculative design.
- **Shared `useSSEStream` hook.** Valuable refactor but not needed until we actually ship SSE.

## Open questions to decide before opening the PR

1. **Old-row shim behavior**: drop phase-level resources when reading legacy JSONB, or attach them all to the first concept? Recommendation: **drop with a log warning**. The bug was that they were duplicated across every concept anyway, so dropping is not a regression.
2. **Question count UX**: fixed at 3 or user-selectable? Recommendation: **fixed at 3 for v1**. Add a dropdown later if users ask.
3. **Regenerate button**: in v1 or in a follow-up? Recommendation: **in v1**. It's one extra button + no new endpoint.

---

### Backend (from repo root)

```bash
cd backend
uv run pytest tests/test_structured_output.py tests/test_assessment_mappers.py \
              tests/test_plan_generator_concepts.py \
              tests/test_concept_generation_route.py tests/test_concept_content_agent.py \
              tests/test_export.py tests/test_state.py \
              tests/test_assessment_service.py tests/test_assessment_routes.py
uv run pytest  # full suite — must be green
```

### OpenAPI sync

```bash
cd backend && python scripts/export-openapi.py
cd frontend && npm run generate-api
# git diff — expect only the intended schema changes; no unrelated drift
```

### Frontend

```bash
cd frontend
npm run typecheck
npm run lint
npm test -- PlanTimeline ConceptCard learning-plan
```

### End-to-end

```bash
cd frontend && npm run test:e2e -- learning-plan
```

### Manual smoke test

1. Start backend + frontend, run an assessment through to completion.
2. Open `/learning-plan`. For a phase with ≥2 concepts, confirm each concept shows a **different** resource list (bug is fixed).
3. On a concept card, click **Study material**. Expect skeleton → inline panel with rendered sections within ~10s.
4. Click **Practice questions**. Expect skeleton → 3 distinct questions with bloom-level badges.
5. Click **Regenerate** on either panel. Expect a fresh response.
6. Load an assessment session that was completed **before this change** (pre-existing JSONB row with old shape). Confirm the plan renders without error. Legacy concepts will render with empty `description` and no resources; that is expected.
7. Check browser DevTools network tab: two new `POST /api/learning-plan/concept/*` requests return JSON < 15s and the response body is camelCase.

---

### Edit in PR 1
- `backend/app/agents/schemas.py` (`PlanResourceOutput` at 49, `PlanPhaseOutput` at 57-67, `PlanOutput` at 70-75)
- `backend/app/graph/state.py` (`Resource` at 119, `LearningPhase` at 125-131, `LearningPlan` at 134-137)
- `backend/app/models/assessment_api.py` (`ResourceOut` at 77, `LearningPhaseOut` at 83-89, `LearningPlanOut` at 92-95)
- `backend/app/agents/plan_generator.py` (`generate_plan` at 14-65; rebuild the phase loop at 45-57)
- `backend/app/prompts/plan_generator.py` (whole file, 15 lines)
- `backend/app/services/assessment_mappers.py` (`build_learning_plan_out` at 85-127)
- `backend/app/routes/export_utils.py` (`build_assessment_markdown` plan section at 86-113 — must edit, not audit)
- `frontend/src/components/learning-plan/PlanTimeline.tsx` (bug at lines 36-42)
- `frontend/src/components/learning-plan/ConceptCard.tsx` (whole file — prop shape change)
- `frontend/src/lib/demo/fixtures.ts` (`DEMO_LEARNING_PLAN` starting line 349)
- `frontend/e2e/fixtures/mock-data.ts` (`learningPlan` starting line 79)

### Create in PR 2
- `backend/app/agents/concept_content.py`
- `backend/app/prompts/concept_content.py`
- `backend/app/models/concept_generation.py`
- `backend/app/routes/concept_generation.py`
- `frontend/src/components/learning-plan/ConceptMaterialPanel.tsx`
- `frontend/src/components/learning-plan/ConceptQuestionsPanel.tsx`
- `backend/tests/test_concept_generation_route.py`
- `backend/tests/test_concept_content_agent.py`
- `frontend/e2e/learning-plan-generation.spec.ts`

### Do NOT touch (deliberately out of scope)
- `backend/app/agents/content_nodes.py` — assessment pipeline content generator, too coupled to `PrioritizedGap` / `TaxonomyIndex` / CLT to reuse; we write new pure functions instead
- `backend/app/agents/question_generator.py` — grew to 264 lines after #167 (adaptive performance signal + evidence sanitization); even deeper coupling to `AssessmentState`. Same reason as above — write new pure functions
- `backend/app/prompts/question_gen.py` — leave the assessment-pipeline prompt alone; write a new prompt for concept questions
- `backend/app/services/sse_adapter.py` — SSE streaming deferred
- `frontend/src/hooks/useAssessmentChat.ts` — SSE streaming deferred

*(Previously listed but now already deleted by #166: `models/learning_plan.py`, `routes/learning_plan.py`, `schemas.py` standalone classes, `PLAN_GENERATOR_SYSTEM_PROMPT`, `ModuleCard.tsx`.)*

### PR DESCRIPTION
**Pull request by onegunsamurai**
## Summary
- **Bug fix**: every concept card rendered the same phase-level resources. Reshaped data model so each concept owns its own `{key, name, description, resources[]}`.
- **Backend**: three-layer schema (`PlanConceptOutput` → `ConceptItem` → `ConceptOut`), JSONB back-compat for legacy rows via `normalize_phase_concepts`, markdown export XSS escaping, `slugify_concept` moved to shared location.
- **Frontend**: `ConceptCard` accepts `ConceptOut`; removed hand-rolled `AssessmentReportResponse` in favor of generated type; a11y improvements (aria-hidden, aria-label, semantic `<ul>/<li>`, stable React keys).

## Quality gates
- 511 backend tests pass (ruff clean)
- 369 frontend tests pass (eslint/tsc clean)
- Pre-commit hooks pass
- Code review, security review, a11y audit findings addressed

## Test plan
- [x] `normalize_phase_concepts`: new shape, legacy `list[str]`, empty, missing key fallback, mixed-type corruption
- [x] `generate_plan`: nested concept output, empty gap nodes
- [x] `build_assessment_markdown`: XSS in concept name, resource URL, knowledge graph HTML injection
- [x] `PlanTimeline.test.tsx`: each concept shows only its own resources (regression)
- [x] `_escape_md` / `_safe_url` unit tests
- [x] Legacy JSONB back-compat for both mapper and markdown export

Closes #168

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: backend/app/services/assessment_mappers.py [hunk 1]
```diff
@@ -82,12 +87,105 @@ def build_gap_analysis_out(enriched) -> GapAnalysis:
...
+
+    Legacy rows lose their phase-level resources on read — those resources
+    were the exact bug that #168 fixes (duplicated across every concept in
+    the frontend), so dropping them is not a regression.
+
+    Shared between ``build_learning_plan_out`` and the markdown exporter so
+    both call sites agree on shape detection.
+    """
+    concepts = phase_dict.get("concepts", [])
+
```
[the remark this answers is not in the document]
Comment (onegunsamurai) [reply backend/app/services/assessment_mappers.py:106]: Fixed in 962f38b — added `isinstance(concepts, list)` guard with warning log. Also added test `test_concepts_not_a_list_returns_empty_and_logs`.

File: backend/app/services/assessment_mappers.py [hunk 2]
```diff
@@ -82,12 +87,105 @@ def build_gap_analysis_out(enriched) -> GapAnalysis:
...
+        else:
+            resources = [r for r in resources if isinstance(r, dict)]
+
+        name = c.get("name", "")
+        return {
+            "key": c.get("key") or slugify_concept(name),
+            "name": name,
+            "description": description,
+            "resources": resources,
+        }
```
[the remark this answers is not in the document]
Comment (onegunsamurai) [reply backend/app/services/assessment_mappers.py]: Fixed in 962f38b — `name` is now coerced via `isinstance(name, str)` check (defaults to ""). Added tests `test_null_name_coerced_to_empty_string` and `test_missing_name_and_key_does_not_crash`.

File: backend/app/routes/export_utils.py [hunk 3]
```diff
@@ -2,6 +2,62 @@
...
+    markdown link early.
+    """
+    if not url:
+        return None
+    url = url.strip()
+    lowered = url.lower()
+    if not (lowered.startswith("http://") or lowered.startswith("https://")):
+        return None
+    # Escape characters that break markdown link syntax.
+    return url.replace("(", "%28").replace(")", "%29")
```
[the remark this answers is not in the document]
Comment (onegunsamurai) [reply backend/app/routes/export_utils.py:56]: Fixed in 962f38b — `_safe_url` now rejects URLs containing any whitespace or control characters (`ord(ch) < 0x20`). Added tests for space, tab, newline, and null byte in URLs.
